### ISSUE TITLE
enable -Wmissing-variable-declarations for W=1

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1920]
I had sent a patch for this, but didn't have time to follow through due to the srso patch set causing a fair amount of breakage for us that week.

https://lore.kernel.org/llvm/20230807-missing_proto-v2-1-3ae2e188bb0c@google.com/

We should work through cleaning up some of the observed issues first before resending the above.

**Comment by JustinStitt** [comment]
A good amount of these are caused by variables with register storage.

https://github.com/llvm/llvm-project/issues/64509 should solve some of these warnings with Clang, right?

**Comment by JustinStitt** [comment]
Can someone help me understand what to do in order to fix some of these warnings?

Here's a case I found from an x86/defconfig build:

```c
arch/x86/ia32/audit.c:6:10: warning: no previous extern declaration for non-static variable 'ia32_dir_class' [-Wmissing-variable-declarations]
    6 | unsigned ia32_dir_class[] = {
```

so it says "no previous extern declaration" so it seems the fix would be to add `static` and call it a day? right?

```c
arch/x86/ia32/audit.c:6:17: warning: unused variable 'ia32_dir_class' [-Wunused-variable]
    6 | static unsigned ia32_dir_class[] = {
      |                 ^~~~~~~~~~~~~~
```

Well now we get an unused variable warning _and_ it seems there **is** an extern:

`$ grep "ia32_dir_class"`

```
...
arch/x86/kernel/audit_64.c
66:     extern __u32 ia32_dir_class[];
73:     audit_register_class(AUDIT_CLASS_DIR_WRITE_32, ia32_dir_class);
...
```

So what am I to make of this? Is the warning wrong? Is there a different solution that I should be going for here?

I wonder if it's caused by the differing types `__u32` vs `unsigned`.


cc @nathanchance 

edit: Ah, it looks like `CONFIG_IA32_EMULATION=y` is required to expose the `extern`.

**Comment by nathanchance** [comment]
It seems like in that case, the declarations in `arch/x86/kernel/audit_64.c` should be hoisted into a header file (perhaps `arch/x86/include/asm/audit.h`?) that is included in both `arch/x86/kernel/audit_64.c` and  `arch/x86/ia32/audit.c`. It also seems like the type of `ia32_dir_class` should be changed from `unsigned` to `__u32`?

**Comment by JustinStitt** [comment]
@nathanchance should they be hiding:
```c
#ifdef CONFIG_IA32_EMULATION
	extern __u32 ia32_dir_class[];
	extern __u32 ia32_write_class[];
	extern __u32 ia32_read_class[];
	extern __u32 ia32_chattr_class[];
	extern __u32 ia32_signal_class[];
	audit_register_class(AUDIT_CLASS_WRITE_32, ia32_write_class);
	audit_register_class(AUDIT_CLASS_READ_32, ia32_read_class);
	audit_register_class(AUDIT_CLASS_DIR_WRITE_32, ia32_dir_class);
	audit_register_class(AUDIT_CLASS_CHATTR_32, ia32_chattr_class);
	audit_register_class(AUDIT_CLASS_SIGNAL_32, ia32_signal_class);
#endif
```

behind this config or have this in a header like you said (without the ifdef)

**Comment by nathanchance** [comment]
I don't think the declarations need to be put under an `#ifdef` in the header file, if that is what you are asking?

**Comment by JustinStitt** [comment]
> I don't think the declarations need to be put under an `#ifdef` in the header file, if that is what you are asking?

Yeah, that's what I'm wodering. Why are they in the ifdef at all. I'll test moving them out.
